File: generate_slip_classifications.py
import random
from datetime import datetime, timedelta

# Define slip classification categories with unique dimensions
slip_categories = [
    # Small Vessels
    {"name": "Compact Runabout", "max_length": 18.0, "max_beam": 7.0, "base_price": 2500},
    {"name": "Small Bowrider", "max_length": 20.0, "max_beam": 7.5, "base_price": 2800},
    {"name": "Small Cuddy Cabin", "max_length": 22.0, "max_beam": 8.0, "base_price": 3200},
    {"name": "Small Cruiser", "max_length": 24.0, "max_beam": 8.5, "base_price": 3600},
    {"name": "Small Sport Boat", "max_length": 25.0, "max_beam": 8.5, "base_price": 3800},
    
    # Medium Vessels
    {"name": "Medium Bowrider", "max_length": 26.0, "max_beam": 9.0, "base_price": 4200},
    {"name": "Medium Express", "max_length": 28.0, "max_beam": 9.5, "base_price": 4800},
    {"name": "Medium Sport Yacht", "max_length": 30.0, "max_beam": 10.0, "base_price": 5600},
    
    # Large Vessels
    {"name": "Large Express Cruiser", "max_length": 33.0, "max_beam": 10.5, "base_price": 6800},
    {"name": "Large Sport Yacht", "max_length": 34.0, "max_beam": 11.0, "base_price": 7200},
    {"name": "Large Cabin Cruiser", "max_length": 35.0, "max_beam": 11.0, "base_price": 7600},
    
    # Extra Large Vessels
    {"name": "Extra Large Express", "max_length": 36.0, "max_beam": 11.5, "base_price": 8200},
    {"name": "Extra Large Sport Yacht", "max_length": 38.0, "max_beam": 12.0, "base_price": 9200},
    {"name": "Extra Large Cabin Yacht", "max_length": 40.0, "max_beam": 12.5, "base_price": 10200},
    
    # Premium Vessels
    {"name": "Premium Motor Yacht", "max_length": 42.0, "max_beam": 13.0, "base_price": 11400},
    {"name": "Premium Sport Yacht", "max_length": 44.0, "max_beam": 13.5, "base_price": 12600},
    
    # Luxury Vessels
    {"name": "Luxury Express Yacht", "max_length": 46.0, "max_beam": 14.0, "base_price": 14000},
    {"name": "Luxury Cabin Yacht", "max_length": 48.0, "max_beam": 14.5, "base_price": 15400},
    {"name": "Luxury Mega Yacht", "max_length": 50.0, "max_beam": 15.0, "base_price": 16800},
    
    # Super Luxury Vessels
    {"name": "Super Luxury Express", "max_length": 52.0, "max_beam": 15.0, "base_price": 18200},
    {"name": "Super Luxury Motor Yacht", "max_length": 54.0, "max_beam": 15.5, "base_price": 19600},
    {"name": "Super Luxury Cabin Yacht", "max_length": 56.0, "max_beam": 16.0, "base_price": 21200},
    {"name": "Super Luxury Mega Yacht", "max_length": 60.0, "max_beam": 16.5, "base_price": 24200},
    
    # Mega Yacht Category
    {"name": "Mega Yacht Standard", "max_length": 65.0, "max_beam": 17.0, "base_price": 27000},
    {"name": "Mega Yacht Deluxe", "max_length": 70.0, "max_beam": 17.5, "base_price": 30200},
    {"name": "Mega Yacht Premium", "max_length": 75.0, "max_beam": 18.0, "base_price": 33600},
    
    # Super Yacht Category
    {"name": "Super Yacht Standard", "max_length": 80.0, "max_beam": 18.5, "base_price": 37500},
    {"name": "Super Yacht Deluxe", "max_length": 85.0, "max_beam": 19.0, "base_price": 41800},
    {"name": "Super Yacht Premium", "max_length": 90.0, "max_beam": 19.5, "base_price": 46500},
    {"name": "Super Yacht Elite", "max_length": 95.0, "max_beam": 20.0, "base_price": 51600},
    {"name": "Super Yacht Ultimate", "max_length": 100.0, "max_beam": 20.5, "base_price": 57200},
    
    # Ultra Luxury Category
    {"name": "Ultra Luxury Standard", "max_length": 110.0, "max_beam": 21.0, "base_price": 64500},
    {"name": "Ultra Luxury Deluxe", "max_length": 120.0, "max_beam": 22.0, "base_price": 72800},
    {"name": "Ultra Luxury Premium", "max_length": 130.0, "max_beam": 23.0, "base_price": 82200},
    {"name": "Ultra Luxury Elite", "max_length": 140.0, "max_beam": 24.0, "base_price": 92800},
    {"name": "Ultra Luxury Ultimate", "max_length": 150.0, "max_beam": 25.0, "base_price": 104500},
    
    # Commercial/Charter Category
    {"name": "Charter Vessel Standard", "max_length": 80.0, "max_beam": 18.0, "base_price": 35000},
    {"name": "Charter Vessel Large", "max_length": 100.0, "max_beam": 20.0, "base_price": 55000}
]

def remove_duplicates(categories):
    """Remove duplicate slip classifications based on unique max_length/max_beam combinations"""
    seen = set()
    unique_categories = []
    
    for category in categories:
        key = (category["max_length"], category["max_beam"])
        if key not in seen:
            seen.add(key)
            unique_categories.append(category)
    
    return unique_categories

def generate_timestamp():
    """Generate a random timestamp between 2020 and 2024 for slip classification creation dates"""
    start_date = datetime(2020, 1, 1)
    end_date = datetime(2024, 12, 31)
    time_between = end_date - start_date
    days_between = time_between.days
    random_days = random.randrange(days_between)
    random_seconds = random.randrange(86400)
    return start_date + timedelta(days=random_days, seconds=random_seconds)

def generate_description(category):
    """Generate a concise description under 80 characters"""
    max_length = category["max_length"]
    max_beam = category["max_beam"]
    price_per_day = category["base_price"] / 100  # Convert cents to dollars
    
    # Create concise description based on vessel size
    if max_length <= 25:
        return f"Compact slip for vessels up to {max_length}'L x {max_beam}'B - ${price_per_day:.0f}/day"
    elif max_length <= 35:
        return f"Medium slip for vessels up to {max_length}'L x {max_beam}'B - ${price_per_day:.0f}/day"
    elif max_length <= 50:
        return f"Large slip for vessels up to {max_length}'L x {max_beam}'B - ${price_per_day:.0f}/day"
    elif max_length <= 75:
        return f"Luxury slip for yachts up to {max_length}'L x {max_beam}'B - ${price_per_day:.0f}/day"
    elif max_length <= 100:
        return f"Super yacht slip up to {max_length}'L x {max_beam}'B - ${price_per_day:.0f}/day"
    else:
        return f"Ultra luxury slip up to {max_length}'L x {max_beam}'B - ${price_per_day:.0f}/day"

# Set seed for reproducible results
random.seed(42)

# Remove duplicates and get unique slip classifications
unique_categories = remove_duplicates(slip_categories)
total_records = len(unique_categories)

# Generate the SQL file
print(f'Generating SlipClassification SQL file with {total_records} unique records...')

with open('slip_classifications.sql', 'w') as f:
    f.write('-- SlipClassification Data Generation Script\n')
    f.write(f'-- Generates {total_records} unique slip classification records\n')
    f.write('-- Based on unique combinations of MaxLength and MaxBeam\n')
    f.write('-- Covers vessel sizes from 18ft runabouts to 150ft+ ultra luxury yachts\n')
    f.write('-- Pricing is based on slip dimensions and amenities\n')
    f.write('-- Base prices are in cents (integer field) for daily rental rates\n\n')
    f.write('BEGIN;\n\n')
    
    f.write('INSERT INTO skipper.slip_classifications ("Name", "MaxLength", "MaxBeam", "BasePrice", "Description", "CreatedAt", "UpdatedAt", "IsDeleted") VALUES\n')
    
    records = []
    for i, category in enumerate(unique_categories):
        name = category["name"]
        max_length = category["max_length"]
        max_beam = category["max_beam"]
        base_price = category["base_price"]
        description = generate_description(category)
        timestamp = generate_timestamp()
        timestamp_str = timestamp.strftime('%Y-%m-%d %H:%M:%S+00')
        
        # Escape single quotes in description for SQL
        description_escaped = description.replace("'", "''")
        
        record = f"('{name}', {max_length}, {max_beam}, {base_price}, '{description_escaped}', '{timestamp_str}', '{timestamp_str}', false)"
        records.append(record)
    
    f.write(',\n'.join(records))
    f.write(';\n\n')
    
    f.write('COMMIT;\n\n')
    f.write('-- Script Summary:\n')
    f.write(f'-- Total Records: {total_records}\n')
    f.write('-- Size Range: 18ft - 150ft vessels\n')
    f.write('-- Beam Range: 7ft - 25ft vessels\n')
    f.write('-- Price Range: $25.00 - $1,045.00 per day\n')
    f.write('-- Categories: Compact, Medium, Large, Luxury, Super Yacht, Ultra Luxury\n')
    f.write('-- All records have IsDeleted = false\n')
    f.write('-- Timestamps distributed between 2020-2024\n')
    f.write('-- Descriptions are concise and under 80 characters\n')
    f.write('-- Only unique MaxLength/MaxBeam combinations included\n')

print(f'SlipClassification SQL file generated: slip_classifications.sql')
print(f'File contains {total_records} unique slip classification records')
print(f'Removed {len(slip_categories) - total_records} duplicate dimension combinations')
print('Descriptions are now concise and under 80 characters') 

File: SkipperDataAgent/GenerateSlipClassifications.cs
using Microsoft.Extensions.Logging;
using SkipperData.Data;
using SkipperData.Data.Repositories;
using SkipperModels;
using SkipperModels.Entities;

namespace SkipperDataAgent;

public static class GenerateSlipClassifications
{
    // Define slip classification categories with unique dimensions
    private static readonly SlipClassificationData[] SlipCategories =
    [
        // Small Vessels
        new("Compact Runabout", 18.0m, 7.0m, 2500),
        new("Small Bowrider", 20.0m, 7.5m, 2800),
        new("Small Cuddy Cabin", 22.0m, 8.0m, 3200),
        new("Small Cruiser", 24.0m, 8.5m, 3600),
        new("Small Sport Boat", 25.0m, 8.5m, 3800),
        
        // Medium Vessels
        new("Medium Bowrider", 26.0m, 9.0m, 4200),
        new("Medium Express", 28.0m, 9.5m, 4800),
        new("Medium Sport Yacht", 30.0m, 10.0m, 5600),
        
        // Large Vessels
        new("Large Express Cruiser", 33.0m, 10.5m, 6800),
        new("Large Sport Yacht", 34.0m, 11.0m, 7200),
        new("Large Cabin Cruiser", 35.0m, 11.0m, 7600),
        
        // Extra Large Vessels
        new("Extra Large Express", 36.0m, 11.5m, 8200),
        new("Extra Large Sport Yacht", 38.0m, 12.0m, 9200),
        new("Extra Large Cabin Yacht", 40.0m, 12.5m, 10200),
        
        // Premium Vessels
        new("Premium Motor Yacht", 42.0m, 13.0m, 11400),
        new("Premium Sport Yacht", 44.0m, 13.5m, 12600),
        
        // Luxury Vessels
        new("Luxury Express Yacht", 46.0m, 14.0m, 14000),
        new("Luxury Cabin Yacht", 48.0m, 14.5m, 15400),
        new("Luxury Mega Yacht", 50.0m, 15.0m, 16800),
        
        // Super Luxury Vessels
        new("Super Luxury Express", 52.0m, 15.0m, 18200),
        new("Super Luxury Motor Yacht", 54.0m, 15.5m, 19600),
        new("Super Luxury Cabin Yacht", 56.0m, 16.0m, 21200),
        new("Super Luxury Mega Yacht", 60.0m, 16.5m, 24200),
        
        // Mega Yacht Category
        new("Mega Yacht Standard", 65.0m, 17.0m, 27000),
        new("Mega Yacht Deluxe", 70.0m, 17.5m, 30200),
        new("Mega Yacht Premium", 75.0m, 18.0m, 33600),
        
        // Super Yacht Category
        new("Super Yacht Standard", 80.0m, 18.5m, 37500),
        new("Super Yacht Deluxe", 85.0m, 19.0m, 41800),
        new("Super Yacht Premium", 90.0m, 19.5m, 46500),
        new("Super Yacht Elite", 95.0m, 20.0m, 51600),
        new("Super Yacht Ultimate", 100.0m, 20.5m, 57200),
        
        // Ultra Luxury Category
        new("Ultra Luxury Standard", 110.0m, 21.0m, 64500),
        new("Ultra Luxury Deluxe", 120.0m, 22.0m, 72800),
        new("Ultra Luxury Premium", 130.0m, 23.0m, 82200),
        new("Ultra Luxury Elite", 140.0m, 24.0m, 92800),
        new("Ultra Luxury Ultimate", 150.0m, 25.0m, 104500),
        
        // Commercial/Charter Category
        new("Charter Vessel Standard", 80.0m, 18.0m, 35000),
        new("Charter Vessel Large", 100.0m, 20.0m, 55000)
    ];

    public static async Task GenerateAsync(SkipperContext dbContext, ILogger<SlipClassificationRepository> logger)
    {
        logger.LogInformation("Starting slip classification generation");
        
        var repository = new SlipClassificationRepository(dbContext, logger);
        var random = new Random(42); // Fixed seed for reproducible results
        
        // Remove duplicates based on unique max_length/max_beam combinations
        var uniqueCategories = RemoveDuplicates(SlipCategories);
        var totalRecords = uniqueCategories.Length;
        
        logger.LogInformation("Generating {TotalRecords} unique slip classifications", totalRecords);
        
        var slipClassifications = new List<SlipClassificationEntity>();
        
        foreach (var category in uniqueCategories)
        {
            var slipClassification = GenerateSlipClassification(category, random);
            slipClassifications.Add(slipClassification);
        }
        
        // Add all slip classifications in parallel
        Task[] classificationTasks = slipClassifications.Select(Task (slipClassification) => repository.AddAsync(slipClassification)).ToArray();
        Task.WaitAll(classificationTasks);
        await repository.SaveChangesAsync();
        
        logger.LogInformation("Slip classification generation completed: {TotalRecords} classifications inserted", totalRecords);
    }

    private static SlipClassificationData[] RemoveDuplicates(SlipClassificationData[] categories)
    {
        var seen = new HashSet<(decimal maxLength, decimal maxBeam)>();
        var uniqueCategories = new List<SlipClassificationData>();
        
        foreach (var category in categories)
        {
            var key = (category.MaxLength, category.MaxBeam);
            if (seen.Add(key))
            {
                uniqueCategories.Add(category);
            }
        }
        
        return uniqueCategories.ToArray();
    }

    private static SlipClassificationEntity GenerateSlipClassification(SlipClassificationData category, Random random)
    {
        var description = GenerateDescription(category);
        var timestamp = GenerateTimestamp(random);

        return new SlipClassificationEntity
        {
            Name = category.Name,
            MaxLength = category.MaxLength,
            MaxBeam = category.MaxBeam,
            BasePrice = category.BasePrice,
            Description = description,
            CreatedAt = timestamp,
            UpdatedAt = timestamp,
            IsDeleted = false
        };
    }

    private static string GenerateDescription(SlipClassificationData category)
    {
        var maxLength = category.MaxLength;
        var maxBeam = category.MaxBeam;
        var pricePerDay = category.BasePrice / 100; // Convert cents to dollars
        
        // Create concise description based on vessel size
        if (maxLength <= 25)
            return $"Compact slip for vessels up to {maxLength}'L x {maxBeam}'B - ${pricePerDay}/day";
        else if (maxLength <= 35)
            return $"Medium slip for vessels up to {maxLength}'L x {maxBeam}'B - ${pricePerDay}/day";
        else if (maxLength <= 50)
            return $"Large slip for vessels up to {maxLength}'L x {maxBeam}'B - ${pricePerDay}/day";
        else if (maxLength <= 75)
            return $"Luxury slip for yachts up to {maxLength}'L x {maxBeam}'B - ${pricePerDay}/day";
        else if (maxLength <= 100)
            return $"Super yacht slip up to {maxLength}'L x {maxBeam}'B - ${pricePerDay}/day";
        else
            return $"Ultra luxury slip up to {maxLength}'L x {maxBeam}'B - ${pricePerDay}/day";
    }

    private static DateTime GenerateTimestamp(Random random)
    {
        var startDate = new DateTime(2020, 1, 1, 0, 0, 0, DateTimeKind.Utc);
        var endDate = new DateTime(2024, 12, 31, 23, 59, 59, DateTimeKind.Utc);
        var timeSpan = endDate - startDate;
        var randomTimeSpan = TimeSpan.FromTicks((long)(random.NextDouble() * timeSpan.Ticks));
        return startDate + randomTimeSpan;
    }

    private record SlipClassificationData(string Name, decimal MaxLength, decimal MaxBeam, int BasePrice);
} 